core/src/com/unciv/logic/map/tile/TileStatFunctions.kt

```diff
@@ -150,51 +171,88 @@ class TileStatFunctions(val tile: Tile) {
 
     // Only gets the tile percentage bonus, not the improvement percentage bonus
     @Suppress("MemberVisibilityCanBePrivate")
-    fun getTilePercentageStats(observingCiv: Civilization?, city: City?, uniqueCache: LocalUniqueCache): Stats {
-        val stats = Stats()
+    fun getTilePercentageStats(observingCiv: Civilization?, city: City?, uniqueCache: LocalUniqueCache): HashMap<String, Stats> {
+        val terrainStats = Stats()
         val stateForConditionals = StateForConditionals(civInfo = observingCiv, city = city, tile = tile)
 
+        val improvement = tile.getUnpillagedTileImprovement()
+        val improvementStats = Stats()
+
+        val road = tile.getUnpillagedRoadImprovement()
+        val roadStats = Stats()
+
         if (city != null) {
             val cachedStatPercentFromObjectCityUniques = uniqueCache.forCityGetMatchingUniques(
                 city, UniqueType.StatPercentFromObject, stateForConditionals)
 
             for (unique in cachedStatPercentFromObjectCityUniques) {
+                val amount = unique.params[0].toFloat()
+                val stat = Stat.valueOf(unique.params[1])
                 val tileFilter = unique.params[2]
-                if (tile.matchesTerrainFilter(tileFilter, observingCiv))
-                    stats[Stat.valueOf(unique.params[1])] += unique.params[0].toFloat()
+
+                if (tile.matchesFilter(tileFilter, observingCiv, true))
+                    terrainStats.add(stat, amount)
+                if (improvement != null && improvement.matchesFilter(tileFilter))
+                    improvementStats.add(stat, amount)
+                if (road != null && road.matchesFilter(tileFilter))
+                    roadStats.add(stat, amount)
             }
 
             val cachedAllStatPercentFromObjectCityUniques = uniqueCache.forCityGetMatchingUniques(
                 city, UniqueType.AllStatsPercentFromObject, stateForConditionals)
             for (unique in cachedAllStatPercentFromObjectCityUniques) {
                 val tileFilter = unique.params[1]
-                if (!tile.matchesTerrainFilter(tileFilter, observingCiv)) continue
                 val statPercentage = unique.params[0].toFloat()
-                for (stat in Stat.values())
-                    stats[stat] += statPercentage
+
+                if (tile.matchesFilter(tileFilter, observingCiv, true))
+                    for (stat in Stat.values())
+                        terrainStats.add(stat, statPercentage)
+                if (improvement != null && improvement.matchesFilter(tileFilter))
+                    for (stat in Stat.values())
+                        improvementStats.add(stat, statPercentage)
+                if (road != null && road.matchesFilter(tileFilter))
+                    for (stat in Stat.values())
+                        roadStats.add(stat, statPercentage)
```

Comment:
There's a lot of repetition here as well, for the 'if tile, if improvement, if road'

How about:
fun addStatPercent(filter, stat, percent):
   if (tile.matches....) terrainStats.add(stat,percent)
   if (improvement...) improvementStats.add(...)
   if (road...) roadStats.add(...) 


At like line 183

Then the 6-line repititions become 1 line, and the 9-line ones become 2
for stat in Stat.values() addStatPercent(filter, stat, statPercentage)